File: bl/json_validate.py
"""bl/json_validate.py — validate the JSON output block in BL 2.0 findings."""

import json
import re

REQUIRED_JSON_FIELDS = {"verdict", "question_id"}

_JSON_FENCE_PATTERN = re.compile(r"```json\s*\n(.*?)```", re.DOTALL)


def validate_finding_json(finding_text: str) -> tuple[dict | None, str | None]:
    """Extract and validate the LAST fenced ```json block in finding_text.

    Returns:
      (parsed_dict, None)      — valid JSON with all required fields present
      (None, error_message)    — JSON block found but malformed or missing fields
      (None, None)             — no JSON block found (prose format, not an error)
    """
    matches = _JSON_FENCE_PATTERN.findall(finding_text)
    if not matches:
        return None, None

    raw = matches[-1].strip()

    try:
        parsed = json.loads(raw)
    except json.JSONDecodeError as exc:
        return None, f"JSON parse error: {exc}"

    missing = REQUIRED_JSON_FIELDS - parsed.keys()
    if missing:
        field_list = ", ".join(sorted(missing))
        return None, f"missing required fields: {field_list}"

    return parsed, None


def is_retry(question_status: str) -> bool:
    """Return True if status contains 'PENDING_RETRY' or 'FORMAT-RETRY'."""
    return "PENDING_RETRY" in question_status or "FORMAT-RETRY" in question_status

File: masonry/src/engine/json-validate.js
'use strict';
// engine/json-validate.js — Validate the JSON output block in BL 2.0 findings.
//
// Port of bl/json_validate.py to Node.js.

const REQUIRED_JSON_FIELDS = new Set(['verdict', 'question_id']);

const _JSON_FENCE_PATTERN = /```json\s*\n(.*?)```/gs;

function validateFindingJson(findingText) {
  const matches = [...findingText.matchAll(_JSON_FENCE_PATTERN)];
  if (matches.length === 0) return [null, null];

  const raw = matches[matches.length - 1][1].trim();

  let parsed;
  try {
    parsed = JSON.parse(raw);
  } catch (exc) {
    return [null, `JSON parse error: ${exc.message}`];
  }

  const keys = new Set(Object.keys(parsed));
  const missing = [...REQUIRED_JSON_FIELDS].filter(f => !keys.has(f));
  if (missing.length > 0) {
    return [null, `missing required fields: ${missing.sort().join(', ')}`];
  }

  return [parsed, null];
}

function isRetry(questionStatus) {
  return questionStatus.includes('PENDING_RETRY') || questionStatus.includes('FORMAT-RETRY');
}

module.exports = { validateFindingJson, isRetry, REQUIRED_JSON_FIELDS };
